### PR TITLE
Feature/rater

### PR DESCRIPTION
**Pull request by jsp-up**

### CODE REVIEW

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-atoms.module.ts [hunk 1]
```diff
@@ -3,14 +3,18 @@ import { CommonModule } from '@angular/common';
...
 
 export * from './kal-input/kal-input.module';
 export * from './kal-checkbox/kal-checkbox.module';
 export * from './kal-icon/kal-icon.module';
 
 const exports = [
   KalCheckboxModule,
-  KalInputModule
+  KalInputModule,
+  KalIconComponent,
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-atoms.module.ts]: il faut mettre ici les modules, pas les composants

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-atoms.module.ts [hunk 2]
```diff
@@ -3,14 +3,18 @@ import { CommonModule } from '@angular/common';
 
 import { KalCheckboxModule } from './kal-checkbox/kal-checkbox.module';
 import { KalInputModule } from './kal-input/kal-input.module';
+import { KalIconComponent } from './kal-icon/kal-icon.component';
+import { KalRaterComponent } from './kal-rater/kal-rater.component';
 
 export * from './kal-input/kal-input.module';
 export * from './kal-checkbox/kal-checkbox.module';
 export * from './kal-icon/kal-icon.module';
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-atoms.module.ts:11]: il manque l'export du KalRaterModule et du KalIconModule

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.module.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { NgModule } from '@angular/core';
+import { CommonModule } from '@angular/common';
+
+import { KalRaterComponent } from './kal-rater.component';
+import { KalIconModule } from 'src/lib/atoms/kal-icon/kal-icon.module';
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.module.ts]: fix import with relative path

File: src/index.html [hunk 4]
```diff
@@ -9,6 +9,8 @@
   <link rel="icon" type="image/x-icon" href="favicon.ico">
 </head>
 <body>
-  <app-root></app-root>
+<app-root></app-root>
+
+<link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
```
Comment (ydomenjoud) [inline src/index.html]: should be in sass, not in index.html

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+   * Notify parent form of the new rate value
+   */
+  rate(rateValue: number): void {
+    this.value = rateValue;
+    super.notifyUpdate(rateValue);
+  }
+
+  ngOnInit() {
+    // create an array from 1 indexed because we can't iterate on numbers in the template
+    this.rateValues = Array.from(new Array(this.maxRate), (val, index) => index + 1);
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts]: Array.fill

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    this.maxRateValue = coerceNumberProperty(rate);
+
+    // update view
+    this.cdr.markForCheck();
+  }
+
+  /**
+   * Name of an icon
+   */
+  @Input()
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts]: input on setter, no on getter

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  private rateValue = 0;
+
+  constructor(private cdr: ChangeDetectorRef) {
+    super();
+  }
+
+  /**
+   * Max rate value
+   */
+  @Input()
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts]: input on setter, no on getter

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+  /**
+   * List of rate values
+   */
+  rateValues: number[];
+
+  /**
+   * Event emitted when the user changed the rate
+   */
+  @Output() rateChanged: EventEmitter<number> = new EventEmitter<number>();
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts]: data is emitted with form control, no need for event emitter

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+@Component({
+  selector: 'kal-rater',
+  templateUrl: './kal-rater.component.html',
+  styleUrls: ['./kal-rater.component.sass'],
+  encapsulation: ViewEncapsulation.None,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  providers: [buildProviders(KalRaterComponent)]
+})
+export class KalRaterComponent extends FormControlAccessComponent implements OnInit {
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts]: should specify generic of FormControlAccessComponent

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+   */
+  @Input()
+  get maxRate(): number {
+    return this.maxRateValue;
+  }
+  set maxRate(rate: number) {
+    this.maxRateValue = coerceNumberProperty(rate);
+
+    // update view
+    this.cdr.markForCheck();
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts]: you should update view on all @Input Change, maybe with an OnChanges interface

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  /**
+   * Name of an icon
+   */
+  @Input()
+  get iconName(): string {
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts]: should inverse iconName with icon ( inverse property name with method name and vice versa ) to simplify input writing

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.html [hunk 12]
```diff
@@ -0,0 +1,5 @@
+<kal-icon *ngFor="let rateValue of rateValues; index as i"
+          [class.active]="value > i"
+          [name]="iconName"
+          (click)="rate(rateValue)">
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.html]: ```rate(index)```

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  }
+
+  /**
+   * Current rate value
+   */
+  get value() {
+    return this.rateValue;
+  }
+  set value(value: number) {
+    this.rateValue = value;
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts]: you should maybe manage value in FormElement class

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.html [hunk 14]
```diff
@@ -0,0 +1,5 @@
+<kal-icon *ngFor="let rateValue of rateValues; index as i"
+          [class.active]="value > i"
```
Comment (ydomenjoud) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.html]: write a method `isActive(index)`, and remove `get value` and `set value`

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts (hunk 8 above)
Comment (ydomenjoud) [reply projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts]: or maybe you should use valueChange of FormElementComponent

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts (hunk 13 above)
Comment (jsp-up) [reply projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts]: I removed the getter and setter for `value`

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.spec.ts [hunk 15]
```diff
@@ -0,0 +1,81 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { KalRaterComponent } from './kal-rater.component';
+import { By } from '@angular/platform-browser';
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.spec.ts]: The angular imports must be above the project imports

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+  templateUrl: './kal-rater.component.html',
+  styleUrls: ['./kal-rater.component.sass'],
+  encapsulation: ViewEncapsulation.None,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  providers: [buildProviders(KalRaterComponent)]
+})
+export class KalRaterComponent extends FormControlAccessComponent<number> implements OnInit, OnChanges {
+
+  /**
+   * List of rate values
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-rater/kal-rater.component.ts]: The number of items in this list corresponds to the number of icons displayed

**Review by ydomenjoud — APPROVED** [review]

**Review by ydomenjoud — APPROVED** [review]

**Comment by ydomenjoud** [comment]
Merge PR KalRater + KalIcon

**Pull request merged**